File: src/vecstore.ts
/**
 * In-process vector store — brute-force cosine similarity over Float32Arrays.
 *
 * Stores vectors in a flat binary file alongside the SQLite vault.
 * Sub-millisecond search for <10k documents. No native extensions needed.
 */

import { existsSync, readFileSync, writeFileSync, mkdirSync } from "node:fs";
import { join, dirname } from "node:path";

interface VectorEntry {
  id: string; // "{content_hash}_{seq}"
  embedding: Float32Array;
}

function cosineSimilarity(a: Float32Array, b: Float32Array): number {
  let dot = 0;
  let normA = 0;
  let normB = 0;
  for (let i = 0; i < a.length; i++) {
    dot += a[i]! * b[i]!;
    normA += a[i]! * a[i]!;
    normB += b[i]! * b[i]!;
  }
  const denom = Math.sqrt(normA) * Math.sqrt(normB);
  return denom === 0 ? 0 : dot / denom;
}

export class VecStore {
  private vectors: Map<string, Float32Array>;
  private filePath: string;
  private dim: number;
  private dirty: boolean;

  constructor(filePath: string, dim = 768) {
    this.filePath = filePath;
    this.dim = dim;
    this.vectors = new Map();
    this.dirty = false;
    this._load();
  }

  /**
   * Add or replace a vector.
   */
  set(id: string, embedding: Float32Array): void {
    this.vectors.set(id, embedding);
    this.dirty = true;
  }

  /**
   * Find the top-k most similar vectors to the query.
   */
  search(query: Float32Array, k = 20): Array<{ id: string; similarity: number }> {
    const results: Array<{ id: string; similarity: number }> = [];

    for (const [id, emb] of this.vectors) {
      results.push({ id, similarity: cosineSimilarity(query, emb) });
    }

    results.sort((a, b) => b.similarity - a.similarity);
    return results.slice(0, k);
  }

  /**
   * Number of stored vectors.
   */
  size(): number {
    return this.vectors.size;
  }

  /**
   * Check if a vector exists.
   */
  has(id: string): boolean {
    return this.vectors.has(id);
  }

  /**
   * Remove a vector.
   */
  delete(id: string): boolean {
    const existed = this.vectors.delete(id);
    if (existed) this.dirty = true;
    return existed;
  }

  /**
   * Persist to disk. Call after batch writes.
   */
  save(): void {
    if (!this.dirty) return;

    const dir = dirname(this.filePath);
    if (!existsSync(dir)) mkdirSync(dir, { recursive: true });

    // Format: [uint32 dim] [uint32 count] then for each entry:
    //   [uint32 id_len] [utf8 id_bytes] [float32[dim] embedding]
    const entries = Array.from(this.vectors.entries());
    const idBuffers = entries.map(([id]) => Buffer.from(id, "utf-8"));

    // Calculate total size
    let totalSize = 8; // dim + count headers
    for (const idBuf of idBuffers) {
      totalSize += 4 + idBuf.length + this.dim * 4;
    }

    const buf = Buffer.alloc(totalSize);
    let offset = 0;

    buf.writeUInt32LE(this.dim, offset);
    offset += 4;
    buf.writeUInt32LE(entries.length, offset);
    offset += 4;

    for (let i = 0; i < entries.length; i++) {
      const [, embedding] = entries[i]!;
      const idBuf = idBuffers[i]!;

      buf.writeUInt32LE(idBuf.length, offset);
      offset += 4;
      idBuf.copy(buf, offset);
      offset += idBuf.length;

      const embBuf = Buffer.from(embedding.buffer, embedding.byteOffset, embedding.byteLength);
      embBuf.copy(buf, offset);
      offset += this.dim * 4;
    }

    writeFileSync(this.filePath, buf);
    this.dirty = false;
  }

  /**
   * Load from disk.
   */
  private _load(): void {
    if (!existsSync(this.filePath)) return;

    try {
      const buf = readFileSync(this.filePath);
      let offset = 0;

      const dim = buf.readUInt32LE(offset);
      offset += 4;
      if (dim !== this.dim) {
        console.error(`Vector dimension mismatch: file has ${dim}, expected ${this.dim}. Ignoring stored vectors.`);
        return;
      }

      const count = buf.readUInt32LE(offset);
      offset += 4;

      for (let i = 0; i < count; i++) {
        const idLen = buf.readUInt32LE(offset);
        offset += 4;
        const id = buf.toString("utf-8", offset, offset + idLen);
        offset += idLen;

        const embedding = new Float32Array(buf.buffer.slice(buf.byteOffset + offset, buf.byteOffset + offset + dim * 4));
        offset += dim * 4;

        this.vectors.set(id, embedding);
      }
    } catch (err) {
      console.error("Failed to load vector store:", err);
    }
  }
}

File: src/mnemon/vecstore.py
"""In-process vector store — brute-force cosine similarity with numpy.

Stores vectors in a .npy file alongside the SQLite vault.
Sub-millisecond search for <10k documents. No native extensions needed.
"""

from __future__ import annotations

import logging
from pathlib import Path

import numpy as np

logger = logging.getLogger(__name__)


class VecStore:
    def __init__(self, file_path: str | Path, dim: int = 384):
        self.file_path = Path(file_path)
        self.dim = dim
        self._ids: list[str] = []
        self._vectors: np.ndarray | None = None  # shape: (n, dim)
        self._dirty = False
        self._load()

    def set(self, vec_id: str, embedding: np.ndarray) -> None:
        """Add or replace a vector."""
        embedding = np.asarray(embedding, dtype=np.float32)
        if embedding.shape != (self.dim,):
            raise ValueError(f"Expected dim {self.dim}, got {embedding.shape}")

        if vec_id in self._ids:
            idx = self._ids.index(vec_id)
            self._vectors[idx] = embedding
        else:
            self._ids.append(vec_id)
            if self._vectors is None:
                self._vectors = embedding.reshape(1, -1)
            else:
                self._vectors = np.vstack([self._vectors, embedding.reshape(1, -1)])
        self._dirty = True

    def search(self, query: np.ndarray, k: int = 20) -> list[dict]:
        """Find the top-k most similar vectors to the query."""
        if self._vectors is None or len(self._ids) == 0:
            return []

        query = np.asarray(query, dtype=np.float32)
        # Cosine similarity: dot(q, v) / (||q|| * ||v||)
        query_norm = np.linalg.norm(query)
        if query_norm == 0:
            return []

        vec_norms = np.linalg.norm(self._vectors, axis=1)
        # Avoid division by zero
        nonzero = vec_norms > 0
        similarities = np.zeros(len(self._ids))
        similarities[nonzero] = (
            self._vectors[nonzero] @ query / (vec_norms[nonzero] * query_norm)
        )

        top_k = min(k, len(self._ids))
        top_indices = np.argpartition(similarities, -top_k)[-top_k:]
        top_indices = top_indices[np.argsort(similarities[top_indices])[::-1]]

        return [
            {"id": self._ids[i], "similarity": float(similarities[i])}
            for i in top_indices
        ]

    def export_all(self) -> tuple[list[str], np.ndarray]:
        """Return all stored (ids, vectors) as a snapshot.

        Used by ``memory_export_vectors`` so the remote dashboard can
        pull the full embedding matrix over MCP and project it with
        UMAP client-side. The returned array is a copy — callers can
        mutate freely without affecting the in-memory store.
        """
        if self._vectors is None or len(self._ids) == 0:
            return [], np.zeros((0, self.dim), dtype=np.float32)
        return list(self._ids), self._vectors.copy()

    def size(self) -> int:
        return len(self._ids)

    def has(self, vec_id: str) -> bool:
        return vec_id in self._ids

    def get(self, vec_id: str) -> np.ndarray | None:
        """Return the vector for ``vec_id``, or ``None`` if not present.

        Returns a defensive copy — callers can mutate freely without
        affecting the in-memory store (matches ``export_all``'s contract).
        """
        if vec_id not in self._ids or self._vectors is None:
            return None
        idx = self._ids.index(vec_id)
        return self._vectors[idx].copy()

    def delete(self, vec_id: str) -> bool:
        if vec_id not in self._ids:
            return False
        idx = self._ids.index(vec_id)
        self._ids.pop(idx)
        if self._vectors is not None:
            self._vectors = np.delete(self._vectors, idx, axis=0)
            if len(self._ids) == 0:
                self._vectors = None
        self._dirty = True
        return True

    def save(self) -> None:
        """Persist to disk."""
        if not self._dirty:
            return
        self.file_path.parent.mkdir(parents=True, exist_ok=True)
        data = {
            "dim": self.dim,
            "ids": self._ids,
            "vectors": self._vectors if self._vectors is not None else np.empty((0, self.dim), dtype=np.float32),
        }
        np.savez(str(self.file_path), **data)
        self._dirty = False

    def _load(self) -> None:
        """Load from disk."""
        npz_path = Path(str(self.file_path) + ".npz") if not str(self.file_path).endswith(".npz") else self.file_path
        if not npz_path.exists():
            return
        try:
            data = np.load(str(npz_path), allow_pickle=True)
            dim = int(data["dim"])
            if dim != self.dim:
                logger.warning(
                    "vecstore: %s has dim=%d, expected %d — ignoring file. "
                    "Run `mnemon rebuild` after a model dimension change.",
                    npz_path, dim, self.dim,
                )
                return
            ids = data["ids"].tolist()
            vectors = data["vectors"]
            if len(ids) > 0 and vectors.shape[0] == len(ids):
                self._ids = ids
                self._vectors = vectors.astype(np.float32)
        except Exception as exc:
            # File exists but is corrupt or unreadable. Silently proceeding
            # with an empty store would mean every memory loses semantic
            # search until the user notices results are degraded — log
            # loudly so a fresh server boot makes the cause obvious.
            logger.warning(
                "vecstore: failed to load %s (%s: %s); starting with empty "
                "in-memory store. Run `mnemon rebuild` to re-embed.",
                npz_path, type(exc).__name__, exc,
            )
